torch/csrc/jit/script/compiler.cpp

```diff
@@ -691,11 +691,56 @@ struct to_ir {
         name, "", std::move(args), std::move(returns), false, false);
   }
 
+  // __setstate__ is special, because it derives the type for `state` from the
+  // output type of __getstate__. This is necessary so that we can allow
+  // submodules to appear in `state`.
+  std::vector<Argument> emitSetStateArguments(
+      const Def& def,
+      const Self* self,
+      const FunctionSchema& schema,
+      Block* block) {
+    std::vector<Argument> arguments; // for schema
+    TORCH_INTERNAL_ASSERT(def.name().name() == "__setstate__");
+    TORCH_INTERNAL_ASSERT(def.decl().params().size() == 2);
+    TORCH_INTERNAL_ASSERT(self);
+    {
+      const auto self_param = def.decl().params()[0];
+      const auto& name = self_param.ident().name();
+      Value* new_input = block->addInput()->setDebugName(name);
+      environment_stack->setSugaredVar(
```

Comment:
I am not a fan of the code duplication here, in particular the environment stack handling. This function should just be providing a type to the argument of  __getstate__ not also duplciating the logic that puts self as a sugared var and initializes the environment. If either of these things change, someone will forget to edit this code.

Reply:
makes sense